### PR TITLE
V4

### PR DESCRIPTION
**Pull request by copleykj**
1. Adds FastRender.onDataReady callback to simplify timing control
2. Adds Extra Data API for sending non pub/sub data with the HTML payload
3. Removes dependency on underscore because modern syntax makes it unnecessary 
4. Improves code structure and refactors to modern syntax, for comprehension and sanity

### CODE REVIEW

**Comment by jankapunkt** [comment]
I have some general suggestions / feedback regarding documentation / code quality

- What do you think about adding a contribution (how to run the linter and tests etc.) section to the README?
- link the `qa/README.md` into the README (or is this file outdated?)
- Running `standard` shows me lint errors (many but only the same kind, `Expected indentation of 2 spaces but found 1 tab`)
- Is the CI still working`I see a `.travis` file in there but can't find it on travis-ci. Maybe move to GitHub CI? I have a working [example repo with Meteor CI on GitHub](https://github.com/leaonline/leaonline-accounts/tree/master/.github/workflows) (for packages this works also great when being [wrapped inside a bare project](https://github.com/serut/meteor-coverage/#specific-setup-for-meteor-package))

What do you think?

**Comment by copleykj** [comment]
@jankapunkt Thanks for the feedback!

I think adding a contribution section, as well as setting up GitHub CI is a great idea. I've removed the travis config file as well as `qa/README.md` as they were presumable artifacts of the original meteorhacks repo and for QA I'd much rather see that fall on tests and not have to do it manually.

As far as linting using the standard command and getting errors, using eslint is probably best since I don't think standard will respect the rules supplied in the eslint config. Honestly though, as far as I'm concerned this config need reworked, getting rid of tabs and adding semicolons, and trailing commas for multiline.

I should be able to start on these changes tonight, and once that's finished, I'll set up husky with a precommit hook that runs eslint.

File: lib/server/publish_context.js [hunk 1]
```diff
@@ -1,118 +1,115 @@
...
-	// our stop does not remove all documents (it just calls deactivate)
-	// Meteor one removes documents for non-universal subscription
-	// we deactivate both for universal and named subscriptions
-	// hopefully this is right in our case
-	// Meteor does it just for named subscriptions
-	this._deactivate()
-}
+PublishContext.prototype.error = function (error) {
+  // TODO: Should we pass the error to the subscription somehow?
+  console.warn(
```
Comment (StorytellerCZ) [inline lib/server/publish_context.js:106]: Maybe instead of console we could use the `logging` package for these.

**Review by StorytellerCZ — CHANGES_REQUESTED** [review]

**Review by StorytellerCZ — APPROVED** [review]

**Comment by copleykj** [comment]
I'm still working to create a solution so that #2 doesn't crop up again. More changes on the way soon.

**Comment by jankapunkt** [comment]
@copleykj yes you can create a bare test project and use `.meteorignore` to keep it out of the published package. I did that with Autoform and since you can then use meteortesting:mocha, chai, sinon and puppeteer you can basically test anything.

**Comment by copleykj** [comment]
ok, so tests are migrated to `meteortesting:mocha` and everything is passing. @jankapunkt will `.meteorignore` actually keep the test app out of the published package? I seem to remember trying to use it to keep `node_modules` out and it failing to do so.

**Comment by jankapunkt** [comment]
@copleykj is there a way to verify that? I based this assumption on the [documentation](https://guide.meteor.com/build-tool.html#compiles-with-build-plugins) and can at least confirm, that it skipped installing the dev dependencies when deploying my app instead of the case where the was no `.meteorignore`

**Comment by copleykj** [comment]
Ok, the testApp directory that I specified in `.meteorignore` **does not** make it into the tarball and so we should be good to go using that method.

**Comment by copleykj** [comment]
It does seem to still include certain directories such as tests, so I think there's a possibility that might have been what was happing with the node_modues dir previously.

Screenshot_2021-05-11_10-24-15

**Comment by copleykj** [comment]
I think things are in pretty good shape at this point. I don't think I'm gonna set up CI right now as I have quite a bit of reading to do and lots to figure out before I'm confident it gets done right.

**Comment by jankapunkt** [comment]
@copleykj GitHub actions take indeed some time to understand when coming from other services like Circle, Travis etc.
I have set up actions for AutoForm so you can peek into it if you need some examples. I can also provide some more complex examples using artifacts, chaining jobs, cloning repos etc., just let me know.

**Pull request merged**